draft-ietf-tls-esni.md

```diff
@@ -801,20 +785,35 @@ of the "ech_required" alert to detect this case.
 
 ### HelloRetryRequest {#server-hrr}
 
-In case a HelloRetryRequest (HRR) is sent, the client-facing server MUST
-consistently accept or decline ECH between the two ClientHellos, using the same
-ECHConfig, and abort the handshake if this is not possible. This is achieved as
-follows. Let CH1 and CH2 denote, respectively, the first and second ClientHello
-transmitted on the wire by the client:
-
-1. If CH1 contains the "encrypted_client_hello" extension but CH2 does not, or
-   if CH2 contains the "encrypted_client_hello" extension but CH1 does not, then
-   the server MUST abort the handshake with an "illegal_parameter" alert.
-1. If the "encrypted_client_hello" extension is sent in CH2, the server follows
-   the procedure in {{client-facing-server}} to decrypt the extension, but it
-   uses the previously-selected ECHConfig as the set of candidate ECHConfigs.
-   If decryption fails, the server aborts the connection with a "decrypt_error"
-   alert rather than continuing the handshake with the second ClientHelloOuter.
+After sending a HelloRetryRequest, the client-facing server does not repeat the
+steps in {{client-facing-server}} with the second ClientHelloOuter. Instead it
+continues the selections from the first ClientHelloOuter as follows:
+
+If the client-facing server accepted ECH, it checks the second ClientHelloOuter
+also contains the "encrypted_client_hello" extension. If not, it MUST abort the
+handshake with a "missing_extension" alert. Otherwise, it decrypts the new
+ClientECH.payload as a second message with the previous HPKE context:
+
+~~~
+    EncodedClientHelloInner = context.Open(ClientHelloOuterAAD,
+                                           ClientECH.payload)
+~~~
+
+ClientHelloOuterAAD is computed as described in {{authenticating-outer}}, but
+using the second ClientHelloOuter. If decryption fails, the client-facing
+server MUST abort the handshake with a "decrypt_error" alert. Otherwise, it
+reconstructs the second ClientHelloInner from the new EncodedClientHelloInner
+as described in {{encoding-inner}}, using the second ClientHelloOuter for
+any referenced extensions.
+
+The client-facing server then forwards the resulting ClientHelloInner to the
+backend server. It forwards all subsequent TLS messages between the client and
+backend server unmodified.
+
+If the client-facing server rejected ECH, or if the first ClientHello did not
+include an "encrypted_client_hello" extension, the client-facing server
+proceeds with the connection as usual, without decrypting any new
```

Comment:
Hrm. I wrote "any" so that this text also works for the no ECH extension case and the "if any" version had lots of clauses and commas. :-)

> If the client-facing server rejected ECH, or if the first ClientHello did not
> include an "encrypted_client_hello" extension, the client-facing server
> proceeds with the connection as usual, without decrypting the new
> ClientECH.payload value, if any.

Given that, are you happy with "any new ClientECH.payload value"? Or do you prefer ", if any" or maybe some other wording altogether?

Reply:
"Any" to me implied "possibly one of the many new values," but the CH only has one value.

Reply:
Hehe. I read "any ClientECH.payload value [singular]" to mean zero or one, but maybe that's not the right way to use that word. :-)

How about:

> If the client-facing server rejected ECH, or if the first ClientHello did not
> include an "encrypted_client_hello" extension, the client-facing server
> proceeds with the connection as usual. The server does not decrypt the
> second ClientHello's ClientECH.payload value, if one was included.

Reply:
Perfect!

Reply:
Reworded.